### PR TITLE
BracesFixer - Fix indentation issues with comments

### LINKED ISSUE #797: Constructor Not Being Indented Correctly
I just ran php-cs-fixer on symfony's 2.6, branch, and found it corrected the indentation on only some functions:

https://github.com/GrahamCampbell/symfony/commit/b25bc38c9898fdf0dec69158938e6ea19a0e26fe#diff-59433ee1eeb1e828c36bb517135f09e5R44

The constructor is incorrectly indented like the other functions, but php-cs-fixer didn't change the indentation.

### LINKED ISSUE #1196: wrong indentation after comment in if statement
All leading whitespaces are tabs.
Fixer is run with psr2 level.
As a result echo '1'; is not indented properly.
If I run fixer with config (listed below) then tab in echo line is not converted and line is not properly indented.

Also fixer doesn't cut to minimum indentation level although I'm not sure if this is a bug or intended behaviour.

.php_cs

``` php
<?php

return Symfony\CS\Config\Config::create()
    ->setUsingCache(true)
    ->level(Symfony\CS\FixerInterface::NONE_LEVEL)
    ->fixers(array(
        'encoding',
        'short_tag',
        'braces',
        'eof_ending',
        'linefeed',
        'parenthesis',
        'php_closing_tag',
        'whitespacy_lines',
        'trailing_spaces',
        'indentation'
    ))
;
```

test.php

``` php
<?php

    if (false) {
    // line after comment is not indented properly
    echo '1';

        $id = 10;
    }
```

### LINKED ISSUE #1344: Indentation fixer -  Wrong indentation
The issue was reported in StyleCI/StyleCI#534.

Since [StyleCI](https://github.com/StyleCI/StyleCI/) uses PHP-CS-Fixer it's better to reference the issue here.

The problem is when the class' code has an indent of 4 spaces and one method has 5 spaces. The `indentation` fixer just fails to fix these lines with 5 spaces.

### LINKED ISSUE #2176: Indentation issue with comment
Hey guys I'm seeing an issue with indentation when an inline comment is present... this has been formated by php-cs-fixer already

```
class fp_charts extends WP_Widget
{
    public function __construct()
    {
        $widget_details = array(
            'classname' => 'fp_charts',
            'description' => 'Create charts using supplied data'
        );
        parent::__construct('fp_charts', 'FlowPress Charts', $widget_details);
    }

    //Backend form
  public function form($instance)
  {
```

### PR DESCRIPTION
**Pull request by julienfalque**
Fixes #797, #1196, #1371 and #2539.
Fixes #2176 at least partially as some comments don't provide input code so I couldn't reproduce them.
May fix #1344, to confirm.
Replaces #2611.

### CODE REVIEW

**Comment by SpacePossum** [comment]
can you add the utests from 2611 (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2611/files) ?

**Comment by julienfalque** [comment]
Tests added. Note that I added them to a different provider method as I don't think they were related to `testPreserveLineAfterControlBrace`. All passed without further changes.

I also added test cases (and a fix) for #2608 which is now fixed by this PR as well. Without this last change, this one was only fixed partially: everything was correcly indented, except multiline content of PHPDoc comments.

**Comment by GrahamCampbell** [comment]
Looks good as far as I can see. 👍

**Comment by julienfalque** [comment]
@GrahamCampbell you reported #797, would it be possible for you to somehow check whether this PR fixes your issue?

**Comment by GrahamCampbell** [comment]
My mac is currently broken, so I am unable to run code easily for now. Taking it in for repair today, so will check on this soon. :)

**Comment by GrahamCampbell** [comment]
Got myself a working Ubuntu environment in the meanwhile. I can confirm that this PR behaves correctly on the issue that I reported. :+1:

**Comment by julienfalque** [comment]
Thanks for checking @GrahamCampbell!

**Comment by SpacePossum** [comment]
```php
    public function provideFixControlContinuationBracesCases()
    {
        return array(
            array(
                '<?php
class A
{
    public function B()
    {
        /*
                */
      $a = 1;
    }
}
',
                '<?php
class A {
    public function B()
    {/*
        */
      $a = 1;
    }
}
',
```

Adding the case shows a problem; after fixing and refixing the code changes.

**Comment by SpacePossum** [comment]
```php
    public function provideFixControlContinuationBracesCases()
    {
        return array(
            array(
                '<?php
class B
{
    public function B()
    {
        /*
            *//**/
        $a = 1;
    }
}
',
                '<?php
class B {
    public function B()
    {
    /*
        *//**/
       $a = 1;
    }
}
',
            ),
            array(
                '<?php
class C
{
    public function C()
    {
        /* */#
        $a = 1;
    }
}
',
                '<?php
class C {
    public function C()
    {
    /* */#
       $a = 1;
    }
}
',
            ),
```

two more cases, not sure if you want to pick up these as well

File: src/Fixer/Basic/BracesFixer.php [hunk 1]
```diff
@@ -429,7 +439,11 @@ function ($item) {
                     || !($nextToken->isWhitespace() && $nextToken->isWhitespace(" \t"))
```
Comment (SpacePossum) [inline src/Fixer/Basic/BracesFixer.php]: while you're at it ;) can you please wrap the `($nextToken [...] && [...] "\n") // preserve blank lines` 
within `(` `)`

and update `// set indent only if it is not a case, when comment is following { in same line` to `on the same line`

**Comment by SpacePossum** [comment]
if you've time, please take a look at the code coverage by the tests as well

File: src/Fixer/Basic/BracesFixer.php (hunk 1 above)
Comment (julienfalque) [inline src/Fixer/Basic/BracesFixer.php]: Updated the comment.

I added braces while fixing the first test case, is it what you wanted to do?

**Comment by julienfalque** [comment]
Test cases added and fixed.

File: src/Fixer/Basic/BracesFixer.php (hunk 1 above)
Comment (SpacePossum) [inline src/Fixer/Basic/BracesFixer.php]: yeah, thanks man :)

**Comment by mooror** [comment]
I'm currently using the latest PHP-CS-Fixer via its atom package and experiencing comment indenting problems myself. I am excited to see a possible fix well under way. 

Keep up the good work,
Mooror

File: src/Fixer/Basic/BracesFixer.php (hunk 1 above)
Comment (SpacePossum) [inline src/Fixer/Basic/BracesFixer.php]: last changes undid the request;
I like when a condition in the form;
```php

if ($a || $b && $c) {
    echo 1;
}
```

is written out explicit as;
```php

if ($a || ($b && $c)) {
    echo 1;
}
```

maybe one day I make a fixer for it ;)

File: src/Fixer/Basic/BracesFixer.php [hunk 2]
```diff
@@ -354,13 +354,35 @@ function ($item) {
...
+                            }
+                        }
+
+                        if ($i === $startBraceIndex) {
+                            $nextLineCanBeIndented = true;
+                        }
+                    }
+                }
+
+                if ($nextLineCanBeIndented) {
```
Comment (SpacePossum) [inline src/Fixer/Basic/BracesFixer.php]: this check can be moved at the end inside the block `if (1 === $nextlevel)` (i.e. remove line `}` at ln 374 and close later)

**Comment by SpacePossum** [comment]
please add;
```php
    public function provideFixControlContinuationBracesCases()
    {
        return array(
            array(
'<?php
if ($a){/*
*/
echo 1;
}
',
```

note, this one is fixed correctly

```php
<?php
if ($a) { /*
*/
echo 1;
}
```

i.e. issue is `{` followed by multi line comment directly

File: src/Fixer/Basic/BracesFixer.php (hunk 2 above)
Comment (julienfalque) [inline src/Fixer/Basic/BracesFixer.php]: Done.

**Comment by julienfalque** [comment]
Rebased onto 2.2 and added the test case.

**Comment by SpacePossum** [comment]
```php
    public function provideFixControlContinuationBracesCases()
    {
        return array(
            array(
'<?php
if ($a) { /**/ /*
*/
    echo 1;
    echo 2;
}
',
                '<?php
if ($a){ /**/ /*
*/
echo 1;
echo 2;
}
',
```

sorry, gave you the wrong case :(

**Comment by julienfalque** [comment]
Tests updated and fixed.

File: src/Fixer/Basic/BracesFixer.php [hunk 3]
```diff
@@ -354,13 +354,47 @@ function ($item) {
...
+                            if (!$tokens[$i]->isWhitespace() && !$tokens[$i]->isComment()) {
+                                break;
+                            }
+                        }
+
+                        if ($i === $startBraceIndex) {
+                            $nextLineCanBeIndented = true;
+                        }
+
+                        if ($nextLineCanBeIndented) {
```
Comment (SpacePossum) [inline src/Fixer/Basic/BracesFixer.php]: can be merged with 375; `if ($i === $startBraceIndex || nextLineCanBeIndented)`

File: src/Fixer/Basic/BracesFixer.php (hunk 1 above)
Comment (SpacePossum) [inline src/Fixer/Basic/BracesFixer.php]:
```php
                if (
                    !$nextNonWhitespaceToken->isComment()
                    || (
                        !($nextToken->isWhitespace() && $nextToken->isWhitespace(" \t"))
                        && ($nextToken->isWhitespace() && 1 === substr_count($nextToken->getContent(), "\n")) // preserve blank lines
                    )
                ) {
```

this test still confuses me haha

File: src/Fixer/Basic/BracesFixer.php (hunk 3 above)
Comment (julienfalque) [inline src/Fixer/Basic/BracesFixer.php]: Done.

Comment (SpacePossum) [inline src/Fixer/Basic/BracesFixer.php]: :+1:

**Comment by SpacePossum** [comment]
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2849#discussion_r125471150
and please throw in some comments to make it easier to follow the code ;)

File: src/Fixer/Basic/BracesFixer.php (hunk 1 above)
Comment (julienfalque) [inline src/Fixer/Basic/BracesFixer.php]: Actually the condition could be simplified, it should be easier to read now :)

**Comment by julienfalque** [comment]
Fixed the first example you provided. I'm not sure about the second one though: how can you tell the indentation of the second comment line is intended and should not be changed?

**Comment by SpacePossum** [comment]
the second comment line is not of concern of the fixer is it? i.e. it is not related to any braces?

**Comment by julienfalque** [comment]
It is: the fixer makes sure the braces are properly placed and that their content is properly indented. Here is the fixer's description:
> The body of each structure MUST be enclosed by braces. Braces should be properly placed. Body of braces should be properly indented.

In the example above, the second comment was not moved because of the various _"previous token is not a comment"_ checks that prevented anything that follows a comment to be properly indented, which is what this PR tries to fix.

**Comment by keradus** [comment]
the second case I provided shall not be modified (and it's not modified on `2.2@stable`)

**Comment by julienfalque** [comment]
> it's not modified on `2.2@stable`

Like I said, anything that follows the first comment is not modified on `2.2@stable`. If you try this:
```php
function foo()
{
    $bar = 1;                   // multiline ...
                                $foo = 3;
    $baz  = 2;                  // next comment
}
```
you'll notice that the `$foo` line isn't modified. With the changes in this PR, it will be, but I see no reason to _not_ fix this line when it is a comment.

**Comment by SpacePossum** [comment]
what about skipping any line that is line break + white space* + comment?

**Comment by julienfalque** [comment]
Then following case would not be changed:
```php
function foo()
{
        // foo
    return $foo;
}
```
which looks wrong to me.

**Comment by keradus** [comment]
the thing is that a lot of users align "single-line" comments to make them be kind-of multiline. breaking it is no-go

**Comment by keradus** [comment]
Extra work made here, glory to you @julienfalque !

**Comment by julienfalque** [comment]
Thank you guys for the help you provided ;)

**Comment by SpacePossum** [comment]
awesome to see this fixed, one of the oldest long standing hard to fix bugs is finally gone :D
thanks @julienfalque !

**Comment by mooror** [comment]
Job well done!

**Comment by MarekLew** [comment]
don`t work corectly for one line comment


```
abstract class XAbcObject
{
    protected $dbc;

    protected $fields; // aa

  protected $emptyFieldsMessages; // bbb
  protected $pk; // cc
  protected $table; // dd
  protected $values = array(); // ee
  protected $errorMsg;  
    protected $errorField;
....
```

**Comment by julienfalque** [comment]
@MarekLew Thanks for your report, can you please open a dedicated issue?

**Pull request closed without merging**